File: src/combinators.ts
import type { Parser } from "./types";
import { Option, Effect } from "effect";
import { success } from "./utils";

export const seq = <A, B>(p1: Parser<A>, p2: Parser<B>): Parser<[A, B]> => ({
  parse: (input) =>
    Effect.flatMap(p1.parse(input), (result1) =>
      Effect.map(p2.parse(result1.remaining), (result2) =>
        success([result1.value, result2.value], result2.remaining)
      )
    ),
});

export const or = <A>(p1: Parser<A>, p2: Parser<A>): Parser<A> => ({
  parse: (input) => Effect.catchAll(p1.parse(input), () => p2.parse(input)),
});

export const map = <A, B>(parser: Parser<A>, fn: (a: A) => B): Parser<B> => ({
  parse: (input) =>
    Effect.map(parser.parse(input), (result) =>
      success(fn(result.value), result.remaining)
    ),
});

export const many = <A>(parser: Parser<A>): Parser<A[]> => ({
  parse: (input) => {
    const results: A[] = [];
    let currentInput = input;

    const parseNext = (): Effect.Effect<never, ParserError, Success<A[]>> =>
      Effect.catchAll(
        Effect.flatMap(parser.parse(currentInput), (result) => {
          results.push(result.value);
          currentInput = result.remaining;
          return parseNext();
        }),
        () => Effect.succeed(success(results, currentInput))
      );

    return parseNext();
  },
});

export const many1 = <A>(parser: Parser<A>): Parser<A[]> => ({
  parse: (input) =>
    Effect.flatMap(parser.parse(input), (firstResult) =>
      Effect.map(many(parser).parse(firstResult.remaining), (restResults) =>
        success(
          [firstResult.value, ...restResults.value],
          restResults.remaining
        )
      )
    ),
});

export const optional = <A>(parser: Parser<A>): Parser<Option.Option<A>> => ({
  parse: (input) =>
    Effect.catchAll(
      Effect.map(parser.parse(input), (result) =>
        success(Option.some(result.value), result.remaining)
      ),
      () => Effect.succeed(success(Option.none(), input))
    ),
});

export const sepBy = <A, S>(
  parser: Parser<A>,
  separator: Parser<S>
): Parser<A[]> => ({
  parse: (input) =>
    Effect.catchAll(
      Effect.flatMap(parser.parse(input), (firstResult) =>
        Effect.map(
          many(map(seq(separator, parser), ([, value]) => value)).parse(
            firstResult.remaining
          ),
          (restResults) =>
            success(
              [firstResult.value, ...restResults.value],
              restResults.remaining
            )
        )
      ),
      () => Effect.succeed(success([], input))
    ),
});

export const between = <L, A, R>(
  left: Parser<L>,
  parser: Parser<A>,
  right: Parser<R>
): Parser<A> => map(seq(seq(left, parser), right), ([[, value]]) => value);

File: parso/combinators.py
from typing import Callable, List
from .types import Parser, Result, Ok, Err, Success
from .utils import success, error
from toolz import curry

def map(parser: Parser, fn: Callable) -> Parser:
    def parse_fn(input):
        result = parser.parse(input)
        if isinstance(result, Ok):
            s = result.value
            return Ok(success(fn(s.value), s.remaining))
        return result
    return Parser(parse_fn)

def flat_map(parser: Parser, fn: Callable) -> Parser:
    def parse_fn(input):
        result = parser.parse(input)
        if isinstance(result, Ok):
            s = result.value
            return fn(s.value).parse(s.remaining)
        return result
    return Parser(parse_fn)

def many(parser: Parser) -> Parser:
    def parse_fn(input):
        results = []
        current_input = input
        while True:
            result = parser.parse(current_input)
            if isinstance(result, Ok):
                s = result.value
                results.append(s.value)
                current_input = s.remaining
            else:
                break
        return Ok(success(results, current_input))
    return Parser(parse_fn)

def optional(parser: Parser) -> Parser:
    def parse_fn(input):
        result = parser.parse(input)
        if isinstance(result, Ok):
            return result
        return Ok(success(None, input))
    return Parser(parse_fn)

def choice(parsers: List[Parser]) -> Parser:
    def parse_fn(input):
        for p in parsers:
            result = p.parse(input)
            if isinstance(result, Ok):
                return result
        return Err(error("No matching choice", input.position))
    return Parser(parse_fn)

def seq(parsers: List[Parser]) -> Parser:
    def parse_fn(input):
        values = []
        current_input = input
        for p in parsers:
            result = p.parse(current_input)
            if isinstance(result, Err):
                return result
            s = result.value
            values.append(s.value)
            current_input = s.remaining
        return Ok(success(values, current_input))
    return Parser(parse_fn)
